crates/sui-adapter/src/adapter.rs

```diff
@@ -126,33 +118,35 @@ pub fn execute<
     args: Vec<CallArg>,
     gas_status: &mut GasStatus,
     ctx: &mut TxContext,
-) -> Result<(), ExecutionError> {
-    let objects = args
-        .iter()
-        .filter_map(|arg| match arg {
-            CallArg::Pure(_) => None,
+) -> Result<Mode::ExecutionResult, ExecutionError> {
+    let mut objects: BTreeMap<ObjectID, &Object> = BTreeMap::new();
+    for arg in &args {
+        match arg {
+            CallArg::Pure(_) => continue,
             CallArg::Object(ObjectArg::ImmOrOwnedObject((id, _, _)))
             | CallArg::Object(ObjectArg::SharedObject { id, .. }) => {
-                Some(vec![(*id, state_view.read_object(id)?)])
+                let obj = state_view.read_object(id);
+                assert_invariant!(obj.is_some(), format!("Object {} does not exist yet", id));
+                objects.insert(*id, obj.unwrap());
             }
-            CallArg::ObjVec(vec) => {
-                if vec.is_empty() {
-                    return None;
+            CallArg::ObjVec(obj_args) => {
+                for obj_arg in obj_args {
+                    match obj_arg {
+                        ObjectArg::ImmOrOwnedObject((id, _, _))
+                        | ObjectArg::SharedObject { id, .. } => {
+                            let obj = state_view.read_object(id);
+                            assert_invariant!(
+                                obj.is_some(),
+                                format!("Object {} does not exist yet", id)
+                            );
+                            objects.insert(*id, obj.unwrap());
+                        }
+                    }
```

Comment:
nit: This is an irrefutable pattern, right?  Could it be turned into a `let` with pattern, to make that clear?

Granted you will need to parenthesise the pattern, so maybe it's not a readability win overall.
```suggestion
                    let (
                        ObjectArg::ImmOrOwnedObject((id, _, _)) | 
                        ObjectArg::SharedObject{ id, .. }
                    ) = obj_arg;
    
                    let obj = state_view.read_object(id);
                    assert_invariant!(
                        obj.is_some(),
                        format!("Object {} does not exist yet", id)
                    );
                    objects.insert(*id, obj.unwrap());
                }
```